glean-core/android/src/main/java/mozilla/telemetry/glean/private/UuidMetricType.kt

```diff
@@ -4,41 +4,74 @@
 
 package mozilla.telemetry.glean.private
 
-import android.util.Log
 import androidx.annotation.VisibleForTesting
-// import mozilla.components.service.glean.Dispatchers
+import com.sun.jna.StringArray
+import mozilla.telemetry.glean.Dispatchers
+import mozilla.telemetry.glean.Glean
+import mozilla.telemetry.glean.rust.getAndConsumeRustString
+import mozilla.telemetry.glean.rust.LibGleanFFI
+import mozilla.telemetry.glean.rust.RustError
+import mozilla.telemetry.glean.rust.toBoolean
+import mozilla.telemetry.glean.rust.toByte
 import java.util.UUID
 
-// import mozilla.components.service.glean.storages.UuidsStorageEngine
-// import mozilla.components.support.base.log.logger.Logger
-
 class UuidMetricType(
-    disabled: Boolean,
-    category: String,
-    lifetime: Lifetime,
-    name: String,
+    private var handle: Long,
     private val sendInPings: List<String>
 ) {
-    companion object {
-        private val LOG_TAG: String = "glean/UuidMetricType"
+    /**
+     * The public constructor used by automatically generated metrics.
+     */
+    constructor(
+        disabled: Boolean,
+        category: String,
+        lifetime: Lifetime,
+        name: String,
+        sendInPings: List<String>
+    ) : this(handle = 0, sendInPings = sendInPings) {
+        val ffiPingsList = StringArray(sendInPings.toTypedArray(), "utf-8")
+        this.handle = LibGleanFFI.INSTANCE.glean_new_uuid_metric(
+            category = category,
+            name = name,
+            send_in_pings = ffiPingsList,
+            send_in_pings_len = sendInPings.size,
+            lifetime = lifetime.ordinal,
+            disabled = disabled.toByte())
+    }
+
+    protected fun finalize() {
+        if (this.handle != 0L) {
+            val error = RustError.ByReference()
+            LibGleanFFI.INSTANCE.glean_destroy_uuid_metric(this.handle, error)
+        }
+    }
+
+    private fun shouldRecord(): Boolean {
+        // Don't record metrics if we aren't initialized
+        if (!Glean.isInitialized()) {
+            return false
+        }
+
+        return LibGleanFFI.INSTANCE.glean_uuid_should_record(Glean.handle, this.handle).toBoolean()
     }
 
     /**
      * Generate a new UUID value and set it in the metric store.
-     *
-     * @return a [UUID] or [null] if we're not allowed to record.
      */
-    fun generateAndSet(): UUID? {
+    fun generateAndSet() {
```

Comment:
We're changing the behaviour of the outer API here. We should always return `UUID?`, since that's our contract with the consumers API. We can discuss about removing this: please file a bug about that.